torchmultimodal/modules/encoders/swin_transformer_3d_encoder.py

```diff
@@ -0,0 +1,512 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+# Modified from 2d Swin Transformers in torchvision:
+# https://github.com/pytorch/vision/blob/main/torchvision/models/swin_transformer.py
```

Comment:
@YosuaMichael I recommend taking the time to create a plan for this and similar models. Not only it will deliver better code quality but also sketch out how we will handle supporting Multimodal more effectively on the future. Copy pasting is a shortcut that kicks down the line the problem and only makes it harder to solve.

Reply:
@ankitade @langong347 Sharing here what we discussed offline with @YosuaMichael and @kartikayk.

The adaptation of Swin 2d to 3d, highlighted some things we can improve on the original implementation of TorchVision. One of them is the use of single integers instead of tuples/lists for the sizes, another one is the fact that some modules can easily be reused if we make minor adaptations (candidates are `PatchMerging` and `SwinTransformerBlock`). It would be really nice, if we could make these changes on TorchVision's code now to minimize the degree of copy-pasting. The reason for the urgency is the upcoming release and the fact that Swin is a brand new class which can be easily modified without worrying about BC now. After the release, there will be all sorts of BC considerations.

The timing is tight and we definitely don't want to delay you. So we will just give it 1-2 days max to see if that's possible. If we are successful to refactor, we will be able to simplify this implementation and reduce the copy-paste. If not, we will merge this and review improvements on TorchVision on the future.

Reply:
Just to update here, the update on torchvision 2d swin transformer is on https://github.com/pytorch/vision/pull/6088